src/main/java/net/minecraftforge/datafix/ForgeSchema.java

```diff
@@ -0,0 +1,340 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.datafix;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.mojang.datafixers.DSL;
+import com.mojang.datafixers.DataFixUtils;
+import com.mojang.datafixers.schemas.Schema;
+import com.mojang.datafixers.types.Type;
+import com.mojang.datafixers.types.families.RecursiveTypeFamily;
+import com.mojang.datafixers.types.families.TypeFamily;
+import com.mojang.datafixers.types.templates.RecursivePoint;
+import com.mojang.datafixers.types.templates.TaggedChoice;
+import com.mojang.datafixers.types.templates.TypeTemplate;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * A partial carbon copy of the {@link Schema} class.
+ * But this allows us to delay the setup of the type hierarchy until we need it.
+ */
+class ForgeSchema extends Schema
+{
+    //Can not be final, needs to be initialized on the fly!
+    //Constructor invocation order and the what not!
+    private Map<String, Supplier<TypeTemplate>> TYPE_TEMPLATES = Maps.newConcurrentMap();
+    private Map<String, Type<?>>                TYPES          = Maps.newConcurrentMap();
+    private Map<String, Integer>                RECURSIVE_TYPES = Maps.newConcurrentMap();
```

Comment:
I believe `RECURSIVE_TYPES` can be final as it isn't assigned anywhere else, but other than that is there a reason why you are redefining `RECURSIVE_TYPES` (and as a `Map<String, Integer>` rather than `Object2IntMap<String>` like in super) when `Schema.RECURSIVE_TYPES` is protected so I believe you could just make direct use of that. Or is it because of how forge's injection works this need to be concurrent?

Reply:
Made it final. It is not a Object2IntMap since those are harder to make properly concurrent (at least I could not find a nice solution) and the performance overhead is not that much of a problem.